### ISSUE TITLE
SameAppKey default value

### ISSUE THREAD

**Issue by jykgod** [googleads/googleads-mobile-unity#2919]
source/plugin/Assets/Plugins/iOS/GADURequestConfiguration.m
```
#import "GADURequestConfiguration.h"
@implementation GADURequestConfiguration
- (instancetype)init {
  self = [super init];
  if (self) {
    _tagForUnderAgeOfConsent = kGADURequestConfigurationTagForUnderAgeOfConsentUnspecified;
    _tagForChildDirectedTreatment =
        kGADURequestConfigurationTagForChildDirectedTreatmentUnspecified;
  }
  return self;
}
@end
```
Why not set the "SameAppKey" default value?
When we use the "RequestConfigurationClient.SetRequestConfiguration" method set the Configuration with SameAppKey=NULL and we found the "SameAppKey" finally be set to false.
It's wrong!We expect the SameAppKey will not to be changed.
Please fix it.

**Comment by NVentimiglia** [comment]
Hi @jykgod 

The SameAppKey on iOS is a non nullable BOOL value:

https://developers.google.com/admob/ios/api/reference/Classes/GADRequestConfiguration#-setsameappkeyenabled

I reason that a default value of `false` is correct. Please let me know If I am misunderstanding.

**Comment by jykgod** [comment]
Hi, thanks for your reply.
I see the Admob document say the SameAppKey is default may be ```true```.

The [Unity Doc](https://developers.google.com/admob/unity/quick-start):
**Same app key**
Prerequisites: Google Mobile Ads Unity Plugin 6.1.0 or higher

The Google Mobile Ads SDK introduces the [same app key](https://support.google.com/admob/answer/9997589) to help you deliver more relevant and personalized ads by using data collected from the app the user is using.

The same app key **is enabled by default**, but you can disable it with the following API:

```
public void Start()
    {
        RequestConfiguration requestConfiguration =
            new RequestConfiguration.Builder()
            .SetSameAppKeyEnabled(true).build();
        MobileAds.SetRequestConfiguration(requestConfiguration);

        // Initialize the Google Mobile Ads SDK.
        MobileAds.Initialize(HandleInitCompleteAction);
    }
```

and The [iOS Doc](https://developers.google.com/admob/ios/global-settings):

**Same app key**
Google Mobile Ads SDK version 8.3.0 introduces the [same app key](https://support.google.com/admanager/answer/9997589) to help you deliver more relevant and personalized ads by using data collected from your apps.

The same app key **is enabled by default**, but you can disable it using the following API.
```
@import GoogleMobileAds

// Disables the same app key.
GADMobileAds.sharedInstance().requestConfiguration.setSameAppKeyEnabled(false)
```

Please confirm it with the relevant department.
This value may affect our income.

**Comment by jykgod** [comment]
Hi @NVentimiglia,
I was wondering if there have been any updates or developments?

**Comment by NVentimiglia** [comment]
@jykgod 

Thanks for the call out, I will get this fixed.

**Comment by NVentimiglia** [comment]
@jykgod 

Hi, I just attempted to replicate this issue and I received a `true` value which is correct. When I did not step the SameAppKeyEnabled value on iOS, loaded an ad, and then called GetRequestConfiguration(), the returned SameAppKeyEnabled value was true as expected.

**Comment by jykgod** [comment]
Hi @NVentimiglia,
You can replicate this issue using the 'new RequestConfiguration. Builder()' or ‘new RequestConfiguration()’ method instead of using the 'GetRequestConfiguration' method. The information we obtained from the document is that “SameAppKey” needs to be true by default, but obtaining the configuration through 'new RequestConfiguration()' is not the case.Just like the code in the   [Targeting Document](https://developers.google.com/admob/unity/targeting):
```
RequestConfiguration requestConfiguration = new RequestConfiguration
{
    TagForChildDirectedTreatment = TagForChildDirectedTreatment.True
};
MobileAds.SetRequestConfiguration(requestConfiguration);
```
If you write the code as above.You can find that the SameAppKey be set to false.

**Comment by NVentimiglia** [comment]
@jykgod 

I see, it looks like the RequestConfiguration on the request is incorrect

I will report the issue, however the `RequestConfiguration.SameAppKey` is actually `true`. What is happening is the `RequestConfiguration` is displaying an incorrect value and the value from `GetRequestConfiguration()` is correct. Please  use `GetRequestConfiguration()` to understand the value of `SameAppKey` for the time being.

**Comment by jykgod** [comment]
@NVentimiglia 
Thank you for your reply again. Perhaps I didn't express myself very clearly. The issue I described is not about not setting the value of SameAppKey or not calling the SetRequestConfiguration method, but about keeping this value empty when setting it, because this value is nullable in Unity SDK as stated in my original description and as previously mentioned in the replies, you can reproduce the issue using the setting method provided in your documentation. If you are still confused, you can use either of the following code snippets to reproduce the problem.
```
RequestConfiguration requestConfiguration = new RequestConfiguration
{
    TagForChildDirectedTreatment = TagForChildDirectedTreatment.True
};
MobileAds.SetRequestConfiguration(requestConfiguration);
Debug("SameAppKey:" + MobileAds.GetRequestConfiguration().SameAppKeyEnabled);
```
or
```
var requestConfiguration = new RequestConfiguration.Builder();
requestConfiguration.TagForChildDirectedTreatment = TagForChildDirectedTreatment.True;
MobileAds.SetRequestConfiguration(requestConfiguration.build());
Debug("SameAppKey:" + MobileAds.GetRequestConfiguration().SameAppKeyEnabled);
```

**Comment by NVentimiglia** [comment]
@jykgod 

I still see the correct values.

```
            MobileAds.SetRequestConfiguration(new RequestConfiguration
            {
                TestDeviceIds = TestDeviceIds
            });
            UnityEngine.Debug.Log("SameAppKey:" + MobileAds.GetRequestConfiguration().SameAppKeyEnabled);
```

![image](https://github.com/googleads/googleads-mobile-unity/assets/125889/5a38482a-0a94-4b54-a95b-5076a86e0f60)

**Comment by jykgod** [comment]
Hi @NVentimiglia,
We used an empty Unity project and imported Admob's Unity SDK. The code above can always be replicated, test result is "false". I checked the SDK code and think it should be due to the SameAppKey value being false in the configuration created through the GADUCreateRequestConfiguration () method in the RequestConfiguration Client file. This is an obvious issue, so I submitted this issue and I am not sure why it cannot be replicated from your end. Do you have any other code before this test code? It is best to delete all other code and only keep this section of code for testing.

**Comment by NVentimiglia** [comment]
@jykgod 

I tested it out again. on iOS, with Unity plugin `8.5.2`. I modified the hello world sample with the following code :

```
            MobileAds.SetRequestConfiguration(new RequestConfiguration
            {
                TestDeviceIds = TestDeviceIds
            });
            UnityEngine.Debug.Log("SameAppKey:" + MobileAds.GetRequestConfiguration().SameAppKeyEnabled);
```

Again, I am seeing that SameAppKeyEnabled is `true`. 

Looking at the code, I see nothing wrong.

For set

[We apply the value](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L59) only if the value is set, preserving the default value.

For Get

The [iOS data type is BOOL](https://developers.google.com/admob/ios/api/reference/Classes/GADRequestConfiguration#-setsameappkeyenabled:)

On [the bridge](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/Plugins/iOS/GADUInterface.m#L758) we read this value as a BOOL

On [the Client](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L78C1-L78C1) we read this value as a BOOL.

**Comment by jykgod** [comment]
@NVentimiglia 
Okay, but the test results on my end are still incorrect. I'm using Admob-Unity-SDK 8.2.0 and conducting the same test on an iOS device with the same code. I can't understand why there would be different results in the same environment. But never mind, I don't want to worry about where the problem is right now. We'll set this value to true explicitly to avoid any issues.

**Comment by jykgod** [comment]
@NVentimiglia 
By the way, although you did not call Externs.GADUSetRequestConfigurationSameAppKeyEnabled to set it when the value of SameAppKey is empty, [Externs.GADUSetRequestConfiguration](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L64) will still be called below to set the entire configuration. The configuration passed in here is created through the [Externs.GADUCreateRequestConfiguration](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L30) function, which is the problematic configuration source I mentioned earlier.
This is unrelated to the read method in Bridge and Client, as a new configuration is created here with the default value of false for SameAppKey.

**Comment by NVentimiglia** [comment]
> The configuration passed in here is created through the [Externs.GADUCreateRequestConfiguration](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L30) function, which is the problematic configuration source I mentioned earlier.

That does looks suspicious. Let me take a deeper look and get back to you.

**Comment by NVentimiglia** [comment]
- The Client Externs [Externs.GADUCreateRequestConfiguration](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L30) calls the ios Bridge.
- The [iOS Bridge](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/Plugins/iOS/GADUInterface.m#L669) handles the command here.
- GADMobileAds.sharedInstance.requestConfiguration is set here, however it is set from the [iOS Request Configuration](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/Plugins/iOS/GADUInterface.m#L627) with a value which was not set. Essentially `sameAppKeyEnabled` = `sameAppKeyEnabled` or `true` = `true` assuming we never mutated the request configuration elsewhere.

If you have some code you would like me to review in C#, let me know.

**Comment by jykgod** [comment]
@NVentimiglia 
The [iOS Request Configuration](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/Plugins/iOS/GADUInterface.m#L627) with a value which was not set.So the value of SameAppKey in the configuration created during [initialization](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/Plugins/iOS/GADURequestConfiguration.m#L3C1-L3C23) here will be the default value of false, right? Then when using this configuration to [set GADMobileAds.sharedInstance.requestConfiguration](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/Plugins/iOS/GADUInterface.m#L669) the requestConfiguration is set to false. This is the logic I have observed and all the code I used for testing has been provided above without any additional C# code. I created a new Unity project as a clean testing environment and my result was false.

**Comment by NVentimiglia** [comment]
> during [initialization](https://github.com/googleads/googleads-mobile-unity/blob/e4aaacca98945f4ce8a9725826a8d67151a1c1f6/source/plugin/Assets/Plugins/iOS/GADURequestConfiguration.m#L3C1-L3C23) here will be the default value of false, right?

No, the default value is `true`.

https://developers.google.com/admob/ios/api/reference/Classes/GADRequestConfiguration#-setsameappkeyenabled:

**Comment by jykgod** [comment]
@NVentimiglia 
Can you pay attention to the question I raised? I am aware that the default value of this parameter in the document should be true. However, I pointed out that there is no initialization of this parameter in the code for creating configurations, which means that its initial value in newly created configurations is false, leading to the issue mentioned above.

**Comment by mmgzentertainment** [comment]
guys, even worse, it doesn't work entirely in android. it is always null.

**Comment by jykgod** [comment]
@mmgzentertainment 
Brother, don't worry about whether this value is null in Android. It's only meant to be used in iOS.

**Comment by mmgzentertainment** [comment]
no brother. I can show you android admob sdk has it. so, this is definitely a bug.
https://developers.google.com/admob/android/global-settings

**Comment by NVentimiglia** [comment]
@mmgzentertainment  I still can not find the bug.

The default value is **true**
https://developers.google.com/admob/ios/api/reference/Classes/GADRequestConfiguration#-setsameappkeyenabled:

We use a nullable type on the API Unity
https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Api/Core/RequestConfiguration.cs;l=75

We only set the value to iOS if we set our value on the API
https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs;l=59

When we read from iOS we use a non-nullable type.
This is a concern...
https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs;l=78

But this is ok, because on iOS we read the value as a non-nullable bool.
https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/Plugins/iOS/GADUInterface.m;l=971

**Comment by jykgod** [comment]
> @mmgzentertainment I still can not find the bug.
> 
> The default value is **true** https://developers.google.com/admob/ios/api/reference/Classes/GADRequestConfiguration#-setsameappkeyenabled:
> 
> We use a nullable type on the API Unity https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Api/Core/RequestConfiguration.cs;l=75
> 
> We only set the value to iOS if we set our value on the API https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs;l=59
> 
> When we read from iOS we use a non-nullable type. This is a concern... https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs;l=78
> 
> But this is ok, because on iOS we read the value as a non-nullable bool. https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/Plugins/iOS/GADUInterface.m;l=971

@NVentimiglia 
1.You create configuration with SameAppKey value is "false":
https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L30

https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/Plugins/iOS/GADUInterface.m#L836

https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/Plugins/iOS/GADURequestConfiguration.m#L4
2.You set the value to iOS without changes:
https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L64

https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/Plugins/iOS/GADUInterface.m#L878

**Comment by NVentimiglia** [comment]
@mmgzentertainment 

> 1.You create configuration with SameAppKey value is "false":
https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L30

> https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/Plugins/iOS/GADUInterface.m#L836

> https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/Plugins/iOS/GADURequestConfiguration.m#L4

For this step we never set the value, which means we are using the [default value](https://developers.google.com/admob/ios/api/reference/Classes/GADRequestConfiguration#-setsameappkeyenabled). 

We only set the value if the publisher explicitly set the value.

https://github.com/googleads/googleads-mobile-unity/blob/main/source/plugin/Assets/GoogleMobileAds/Platforms/iOS/RequestConfigurationClient.cs#L59

That said, I am going to send this issue to another engineer to take a look. I was unable to replicate the issue or identify the problem code, but I want to give this a second opinion.

**Comment by NVentimiglia** [comment]
@jykgod 

I was able to identify an issue with the plugin. The SameAppKey should be write only, and reading from the object may not be correct. I have submitted a request to fix this. Thank you for sticking with me.

**Comment by kunny** [comment]
We're going to introduce a new API that replaces the `SameAppKeyEnabled` in the upcoming release.

For the time being, explicitly set the `SameAppKeyEnabled` value to `true` as a workaround.

```csharp
RequestConfiguration requestConfiguration = new RequestConfiguration
{
    SameAppKeyEnabled = true,
};
MobileAds.SetRequestConfiguration(requestConfiguration);
```

**Comment by NVentimiglia** [comment]
This issue should not be fixed b/308893640
